File: src/lib/fleet.js
import { equipmentTypes } from '../data/demandHistory'
import { utilizationOf } from './rules'

// Per-category availability rollup — drives the dashboard's category table.
export function categorySummary(equipment) {
  return equipmentTypes
    .map((type) => {
      const units = equipment.filter((e) => e.type === type)
      const available = units.filter((e) => e.status === 'completed').length
      const rented = units.filter((e) => e.status === 'active').length
      const maintenance = units.filter((e) => e.status === 'maintenance').length
      const total = units.length
      return {
        type,
        total,
        available,
        rented,
        maintenance,
        availPct: total === 0 ? 0 : Math.round((available / total) * 100),
      }
    })
    .filter((row) => row.total > 0)
}

// Fleet-wide utilization: share of engine hours out of all logged hours
// across everything currently on rent.
export function fleetUtilization(activeUnits) {
  const engine = activeUnits.reduce((s, e) => s + e.avgEngineHoursPerDay, 0)
  const idle = activeUnits.reduce((s, e) => s + e.avgIdleHoursPerDay, 0)
  if (engine + idle === 0) return 0
  return Math.round((engine / (engine + idle)) * 100)
}

// Ranked utilization — top N performers and bottom N, for the leaderboard.
export function utilizationRanking(activeUnits, n = 5) {
  const ranked = [...activeUnits]
    .map((e) => ({ eq: e, util: Math.round(utilizationOf(e) * 100) }))
    .sort((a, b) => b.util - a.util)
  return { top: ranked.slice(0, n), bottom: ranked.slice(-n).reverse() }
}

File: backend/app/services/fleet.py
"""Python port of src/lib/fleet.js — dashboard rollups."""

from .rules import utilization_of

EQUIPMENT_TYPES = ["Excavator", "Bulldozer", "Crane", "Grader", "Forklift", "Loader", "Roller"]


def category_summary(equipment: list[dict]) -> list[dict]:
    rows = []
    for etype in EQUIPMENT_TYPES:
        units = [e for e in equipment if e["type"] == etype]
        total = len(units)
        if total == 0:
            continue
        available = sum(1 for e in units if e["status"] == "completed")
        rows.append(
            {
                "type": etype,
                "total": total,
                "available": available,
                "rented": sum(1 for e in units if e["status"] == "active"),
                "reserved": sum(1 for e in units if e["status"] == "hold"),
                "maintenance": sum(1 for e in units if e["status"] == "maintenance"),
                "availPct": round((available / total) * 100) if total else 0,
            }
        )
    return rows


def fleet_utilization(active_units: list[dict]) -> int:
    engine = sum(e["avg_engine_hours_per_day"] for e in active_units)
    idle = sum(e["avg_idle_hours_per_day"] for e in active_units)
    if engine + idle == 0:
        return 0
    return round((engine / (engine + idle)) * 100)


def time_utilization(equipment: list[dict]) -> int:
    rentable = [e for e in equipment if e["status"] != "maintenance"]
    if not rentable:
        return 0
    on_rent = sum(1 for e in equipment if e["status"] == "active")
    return round((on_rent / len(rentable)) * 100)


def utilization_ranking(active_units: list[dict], n: int = 5) -> dict:
    ranked = sorted(
        ({"eq": e, "util": round(utilization_of(e) * 100)} for e in active_units),
        key=lambda r: r["util"],
        reverse=True,
    )
    return {"top": ranked[:n], "bottom": list(reversed(ranked[-n:])) if ranked else []}
